pkg/redirectpolicy/redirectpolicy.go

```diff
@@ -0,0 +1,180 @@
+	frontend         *loadbalancer.L3n4Addr
+	serviceID        *k8s.ServiceID
+	svcPortProto     *api.PortProtocol
+	backendSelector  *api.EndpointSelector
+	backendPortProto *api.PortProtocol
+	backends         []loadbalancer.L3n4Addr
+}
+
+// Parse parses the specified cilium redirect local policy spec, and returns
+// a sanitized LRPConfig.
+func Parse(clrp *v2.CiliumLocalRedirectPolicy, sanitize bool) (LRPConfig, error) {
+	name := clrp.ObjectMeta.Name
+	if name == "" {
+		return LRPConfig{},
+			fmt.Errorf("CiliumLocalRedirectPolicy must have a name")
+	}
+
+	namespace := k8sUtils.ExtractNamespace(&clrp.ObjectMeta)
+	// if namespace == "" {
+	// TODO CCLRP
+	// 	return cclrp.Parse()
+	// }
+
+	if sanitize {
+		return getSanitizedLRPConfig(name, namespace, clrp.UID, clrp.Spec)
+	} else {
+		return LRPConfig{
+			name:      name,
+			namespace: namespace,
+		}, nil
+	}
+}
+
+func getSanitizedLRPConfig(name, namespace string, uid types.UID,
+	spec v2.CiliumLocalRedirectPolicySpec) (LRPConfig, error) {
+
+	var (
+		addrMatcher  = spec.RedirectFrom.AddressMatcher
+		svcMatcher   = spec.RedirectFrom.ServiceMatcher
+		redirectTo   = spec.RedirectTo
+		k8sSvc       *k8s.ServiceID
+		svcPortProto api.PortProtocol
+		bePortProto  api.PortProtocol
+		fe           *loadbalancer.L3n4Addr
+	)
+
+	if addrMatcher == nil && svcMatcher == nil {
+		return LRPConfig{}, fmt.Errorf("both address and service" +
+			" matchers can not be empty")
+	}
+
+	if addrMatcher != nil && svcMatcher != nil {
+		return LRPConfig{}, fmt.Errorf("both address and service" +
+			" matchers can not be specified")
+	}
+
+	if addrMatcher != nil {
+		// LRP specifies IP and L4 port tuple config for traffic that needs to be redirected.
+		ip := net.ParseIP(addrMatcher.IP)
+		if ip == nil {
+			return LRPConfig{}, fmt.Errorf("invalid address matcher IP %s",
+				addrMatcher.IP)
+		}
+		err := addrMatcher.PortProtocol.Sanitize()
+		if err != nil {
+			return LRPConfig{}, fmt.Errorf("invalid address matcher L4 port"+
+				" protocol %s", err)
+		}
+		p, _ := strconv.ParseUint(addrMatcher.PortProtocol.Port, 0, 16)
+		port := uint16(p)
+		protocol, _ := loadbalancer.NewL4Type(string(addrMatcher.PortProtocol.Protocol))
+		fe = loadbalancer.NewL3n4Addr(protocol, ip, port, loadbalancer.ScopeExternal)
+	} else {
+		if svcMatcher.Service == nil || svcMatcher.Service.K8sService == nil {
+			return LRPConfig{}, fmt.Errorf("empty service matcher" +
+				" kubernetes service")
+		}
+		k8sSvc = &k8s.ServiceID{
+			Name:      svcMatcher.Service.K8sService.ServiceName,
+			Namespace: svcMatcher.Service.K8sService.Namespace,
+		}
+		// LRP specifies service config for traffic that needs to be redirected.
+		// LRP specifies K8s service config for traffic that needs to be redirected.
+		if k8sSvc.Name == "" {
+			return LRPConfig{}, fmt.Errorf("kubernetes service name can" +
+				"not be empty")
+		}
+		if k8sSvc.Namespace == "" {
+			// TODO Cluster scoped
+		}
+		err := svcMatcher.PortProtocol.Sanitize()
```

Comment:
At a glance I couldn't figure out where this `Sanitize()` implementation is supposed to be occurring but I'm not sure it's doing what we think it's doing. Might be a candidate for a unit test.

Reply:
@aditighag pointed out to me that this supports named ports which address my untold assumption here that providing a port like "tcp" can be valid.